aws/aws-app-mesh-controller-for-k8s
# waitUntilProxyReady feature

*Issue #, if available:*

https://github.com/aws/aws-app-mesh-controller-for-k8s/issues/489

*Description of changes:*

Enable pod postStart hook to delay application startup until proxy is ready to accept traffic

Gated with a feature flag:
1) ensure envoy container is first in pod lifecycle
2) add a postStart hook that polls envoy server is ready

Integration spec:
1) new suite to validate behavior
2) includes docker images for frontend and backend apps used in test

Testing instructions:

1) pull branch locally
2) build controller image
```
TARGETPLATFORM=linux/amd64 docker build -t amazon/appmesh-controller .
```
3) tag controller image version
```
docker tag amazon/appmesh-controller:latest $AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/amazon/appmesh-controller:sidecar-init-v1.0.10
```
4) push to ecr repository
```
5) docker push $AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/amazon/appmesh-controller:sidecar-init-v1.0.10
```
6) from repo root install
```
helm upgrade appmesh-controller ./config/helm/appmesh-controller \
    --namespace appmesh-system \
    --set region=$AWS_REGION \
    --set serviceAccount.create=false \
    --set serviceAccount.name=appmesh-controller \
    --set image.repository=$AWS_ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com/amazon/appmesh-controller
    --set image.tag=sidecar-init-v1.0.10
    --set sidecar.waitUntilProxyReady=true
```
7) run the integration test:
```
ginkgo -v -r test/integration/sidecar -- --cluster-kubeconfig=${dir} --cluster-name=${name} --aws-region=$AWS_REGION --aws-vpc-id=${vpc}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
